x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java

```diff
@@ -989,10 +990,12 @@ private AbstractRunnable fillGapRunnable(SparseFileTracker.Gap gap, RangeMissing
                     start,
                     start,
                     Math.toIntExact(gap.end() - start),
-                    progress -> gap.onProgress(start + progress)
+                    progress -> gap.onProgress(start + progress),
+                    ActionListener.wrap(unused -> {
+                        writeCount.increment();
```

Comment:
As it is right now, after the fillCacheRange has been called, the ActionRunnable.run() completes is listener immediately, right? While I'd expect we'd complete here (inside the listener of fillCacheRange) the outside `listener`. Since fillCacheRange may be async.

Reply:
Yes, the idea is that all `fillCacheRange` implementation would become async by default and would need to explicitly call `completed.onResponse(null)`

Reply:
Sure. What I'm saying that instead of  completing the listener in the 
`return ActionRunnable.run(listener.delegateResponse((l, e) -> failGapAndListener(gap, l, e)), () -> {`
above (which happens at the end of run() even if the async thread is still ongoing),
we should remove it from there, and complete the `listener` here inside the .wrap() (i.e., when the async thread completes the `completed` listener).

Reply:
Thank you! That part slipped from me, I assumed it existed purely for error propagation. I've changed it to 
```java
listener.<Void>map(unused -> {
        writeCount.increment();
        gap.onCompletion();
        return null;
}).delegateResponse((l, e) -> failGapAndListener(gap, l, e))
```

Reply:
Can also be done as
```java
ActionListener.wrap(unused -> ActionListener.run(listener, l -> {
          writeCount.increment();
          gap.onCompletion();
          l.onResponse(null);
}), e -> failGapAndListener(gap, listener, e))
```

Not sure which option is better to understand.